types/src/state_store/state_key.rs

```diff
@@ -24,43 +25,51 @@ pub enum StateKey {
     Raw(Vec<u8>),
 }
 
-#[derive(ToPrimitive)]
+#[repr(u8)]
+#[derive(FromPrimitive, ToPrimitive)]
 enum StateKeyPrefix {
     AccountAddress,
     AccessPath,
     Raw = 255,
 }
 
-impl StateKeyPrefix {
-    fn to_bytes(&self) -> Vec<u8> {
-        let byte = self
-            .to_u8()
-            .expect("Failed to convert StateKeyPrefix to u8");
-        vec![byte]
-    }
-}
-
-struct RawStateKey {
-    bytes: Vec<u8>,
-}
+impl StateKey {
+    /// Serializes to bytes for physical storage.
+    pub fn encode(&self) -> anyhow::Result<Vec<u8>> {
+        let mut out = vec![];
 
-impl From<&StateKey> for RawStateKey {
-    fn from(key: &StateKey) -> Self {
-        let (prefix, raw_key) = match key {
-            StateKey::AccountAddressKey(account_address) => {
-                (StateKeyPrefix::AccountAddress, account_address.to_vec())
-            }
+        let (prefix, raw_key) = match self {
+            StateKey::AccountAddressKey(account_address) => (
+                StateKeyPrefix::AccountAddress,
+                bcs::to_bytes(account_address)?,
+            ),
             StateKey::AccessPath(access_path) => {
-                let mut raw_key = access_path.address.to_vec();
-                raw_key.extend(access_path.path.clone());
-                (StateKeyPrefix::AccessPath, raw_key)
+                (StateKeyPrefix::AccessPath, bcs::to_bytes(access_path)?)
             }
             StateKey::Raw(raw_bytes) => (StateKeyPrefix::Raw, raw_bytes.to_vec()),
         };
-        let mut bytes = prefix.to_bytes();
-        bytes.extend(raw_key);
+        out.push(prefix as u8);
+        out.extend(raw_key);
+        Ok(out)
+    }
 
-        Self { bytes }
+    /// Recovers from serialized bytes in physical storage.
+    pub fn decode(val: &[u8]) -> anyhow::Result<StateKey> {
+        if val.is_empty() {
+            return Err(StateKeyDecodeErr::EmptyInput.into());
+        }
+        let tag = val[0];
+        let state_key_tag = StateKeyPrefix::from_u8(tag);
```

Comment:
`.ok_or_else(|| StateKeyDecodeErr::UnknownTag{ .. }?` and we don't need to repeat all the `Some` in the match below.